pyvrp/cpp/ProblemData.cpp

```diff
@@ -42,19 +59,52 @@ ProblemData::Client::Client(Coordinate x,
         throw std::invalid_argument("prize must be >= 0.");
 }
 
+ProblemData::Client::Client(Client const &client)
+    : x(client.x),
+      y(client.y),
+      demand(client.demand),
+      serviceDuration(client.serviceDuration),
+      twEarly(client.twEarly),
+      twLate(client.twLate),
+      releaseTime(client.releaseTime),
+      prize(client.prize),
+      name(duplicate(client.name)),
+      required(client.required)
+{
+}
+
+ProblemData::Client::Client(Client &&client)
+    : x(client.x),
+      y(client.y),
+      demand(client.demand),
+      serviceDuration(client.serviceDuration),
+      twEarly(client.twEarly),
+      twLate(client.twLate),
+      releaseTime(client.releaseTime),
+      prize(client.prize),
+      name(client.name),  // we can steal
+      required(client.required)
+{
+    client.name = nullptr;  // stolen
+}
+
+ProblemData::Client::~Client() { delete[] name; }
+
 ProblemData::VehicleType::VehicleType(size_t numAvailable,
                                       Load capacity,
                                       size_t depot,
                                       Cost fixedCost,
                                       std::optional<Duration> twEarly,
                                       std::optional<Duration> twLate,
-                                      std::optional<Duration> maxDuration)
+                                      std::optional<Duration> maxDuration,
+                                      char const *name)
     : numAvailable(numAvailable),
       depot(depot),
       capacity(capacity),
       fixedCost(fixedCost),
       twEarly(twEarly),
       twLate(twLate),
+      name(duplicate(name)),
```

Comment:
Same here, why is the order of name and max duration reversed?

Reply:
[same holds for Client]

The fields are initialised in the order of their layout in the header file. That layout is:
```cpp
      [8] size_t const numAvailable;  // Available vehicles of this type
      [8] size_t const depot;         // Departure and return depot location
      [4] Load const capacity;        // This type's vehicle capacity
      [4] Cost const fixedCost;       // Fixed cost of using this vehicle type
      [8] std::optional<Duration> const twEarly;  // Start of shift
      [8] std::optional<Duration> const twLate;   // End of shift
      [8] char const *name;                       // Type name (for reference)
      [4] Duration const maxDuration;             // Maximum route duration
```
which is slightly different from argument order in the constructor, which we try to keep backwards compatible.

Each of the items here has a particular size: Cost, Load, Duration, etc. all boil down to `int`, which is 4 bytes on most of the systems we work with. A `size_t` and a pointer type are typically 8 bytes, as is `std::optional`. I've put these sizes in brackets above.

The size of each field determines [alignment requirements](https://en.cppreference.com/w/c/language/object#Alignment). The compiler needs to respect these, and will insert some artificial padding to make that happen. The VehicleType struct now has size `52` in pure data, but `56` in effective size due to a little padding at the end, after `maxDuration`. We can thus add another integer value effectively for 'free' when we need to track more data for future features - adding that to the end now will do the right thing.

Suppose we had done this instead:
```cpp
      [4] Duration const maxDuration;             // Maximum route duration
      [4 bytes padding to align name]
      [8] char const *name;                       // Type name (for reference)
``` 
Then the articial padding would have been 'hidden' between `maxDuration` and `name`, and our struct would have been bigger than it needs to be. 

Since this is all internal we get to change this around a bit whenever we need to. I sometimes tweak it a bit to shave off some bytes.